paddle/fluid/framework/details/nccl_op_handle.h

```diff
@@ -0,0 +1,230 @@
+      auto ctxs = nccl_ctxs_->GetFlatCtx(run_order);
+      for (auto& p : places_) {
+        this->SetDeviceContext(p, ctxs->DevCtx(p));
+      }
+      return;
+    }
+
+    PADDLE_ENFORCE(places_.size() == 1,
+                   "HierarchicalAllReduce run one proc with one card mode.");
+
+    for (auto& p : places_) {
+      auto ctxs = nccl_ctxs_->GetHierarchicalInterCtx(run_order);
+      this->SetDeviceContext(p, ctxs->DevCtx(p));
+    }
+
+    for (auto& p : dev_ctxes_) {
+      int dev_id = boost::get<platform::CUDAPlace>(p.first).device;
+      if (inter_events_.find(dev_id) != inter_events_.end()) {
+        continue;
+      }
+
+      PADDLE_ENFORCE(cudaSetDevice(dev_id));
+      PADDLE_ENFORCE(cudaEventCreateWithFlags(&inter_events_[dev_id],
+                                              cudaEventDisableTiming));
+      PADDLE_ENFORCE(cudaEventCreateWithFlags(&exter_events_[dev_id],
+                                              cudaEventDisableTiming));
+      VLOG(10) << "Create events on dev_id:" << dev_id
+               << ", inter_event:" << &inter_events_[dev_id]
+               << ", exter_event:" << &exter_events_[dev_id];
+    }
+  }
+
+  void FlatNcclAllReduce(platform::Place place, const void* sendbuff,
+                         void* recvbuff, size_t count, ncclDataType_t datatype,
+                         ncclRedOp_t op) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
+    auto flat_nccl_ctxs = nccl_ctxs_->GetFlatCtx(run_order_);
+    int dev_id = boost::get<platform::CUDAPlace>(place).device;
+    auto& nccl_ctx = flat_nccl_ctxs->at(dev_id);
+    auto stream = nccl_ctx.stream();
+    auto comm = nccl_ctx.comm_;
+
+    VLOG(10) << "before all reduce buffer:" << sendbuff << ", numel:" << count
+             << ", dev_id:" << dev_id << ", dtype:" << datatype
+             << ", place:" << place;
+
+    PADDLE_ENFORCE(platform::dynload::ncclAllReduce(
+        sendbuff, recvbuff, count, datatype, op, comm, stream));
+  }
+
+  void NCCLAllReduce(platform::Place place, const void* sendbuff,
+                     void* recvbuff, size_t count, ncclDataType_t datatype,
+                     ncclRedOp_t op) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
+    if (!use_hierarchical_allreduce_) {
+      FlatNcclAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+      return;
+    }
+
+    HierarchicalAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+  }
+
+  void HierarchicalAllReduce(platform::Place place, const void* sendbuff,
+                             void* recvbuff, size_t count,
+                             ncclDataType_t datatype, ncclRedOp_t op) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
+    InterAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+    if (nccl_ctxs_->NeedExterAllReduce()) {
+      ExterAllReduce(place, recvbuff, recvbuff, count, datatype, op);
+    }
+    InterBroadCast(place, recvbuff, count, datatype, op);
```

Comment:
Why do you call `InterBroadCast ` here?

Reply:
I see.

Reply:
If the `nccl_ctxs_->NeedExterAllReduce()` if false, not need to call `InterBroadCast `.

Reply:
We need.
Some trainer need not exter allreduce because there rank in exter allreduce is not 0, but when broadcast they need call broadcast all.

Reply:
I think it will be better:
```
if (nccl_ctxs_->NeedExterAllReduce()) {
 InterReduce(place, sendbuff, recvbuff, count, datatype, op);
 ExterAllReduce(place, recvbuff, recvbuff, count, datatype, op);
 InterBroadCast(place, recvbuff, count, datatype, op);
} else {
 InterAllReduce(place, sendbuff, recvbuff, count, datatype, op);
}
```

Reply:
You missed it. Talk about it later.